### ISSUE TITLE
Request to Expose AudioBuffer to DedicatedWorker

### ISSUE THREAD

**Issue by chcunningham** [WebAudio/web-audio-api-v2#111]
WebCodecs uses AudioBuffer as a container of unencoded audio.
https://wicg.github.io/web-codecs/#dom-audioframe-buffer

The AudioFrame, which wraps AudioBuffer, is provided as an input to AudioEncoder and as an output from AudioDecoder. We've exposed the encoder and decoder interfaces to DedicatedWorker to facilitate offloaded codec IO.

I think the change is as simple as changing the Exposed= parameter for just AudioBuffer. I'll send a PR to demonstrate.

**Comment by padenot** [comment]
(Moving this to V2 because we're not changing V1 text at this stage).

**Comment by hoch** [comment]
For reference: crbug.com/1160580

**Comment by rtoy** [comment]
I would like to know what the actual use case is and whether AudioBuffer is the right api, as @padenot mentions.  Perhaps there's a better solution.

**Comment by chcunningham** [comment]
AudioBuffer is used by WebCodecs to describe raw audio. It is essentially the output of AudioDecoder and the input to AudioEncoder. In both cases, we wrap the buffer in an [AudioFrame](https://wicg.github.io/web-codecs/#audioframe-interface) to add some metadata and the ability to immediately release the buffer without waiting for GC (more important for VideoFrame, but consistency is good).

We like AudioBuffer for several reasons
* it describes a buffer of raw audio in much the same way as other codec apis (e.g. chrome's internal api, ffmpeg, etc...), including sampleRate, length, duration, channels
* it offers mechanisms to operate on the raw float data
* easy integration with WebAudio, including AudioWorklet which will be used for rendering by WebCodecs applications

**Comment by guest271314** [comment]
The underlying `Float32Array` `buffer` can be transferred to `Worker` or `AudioWorkletGlobalScope` using `postMessage(f32, [f32.buffer])`. We can also write data to a `SharedArrayBuffer`. On Chrome and Chromium we can use Transferable Streams to read and write across contexts using a `TransforStream`. Examples using each https://github.com/guest271314/AudioWorkletStream. Do we really need `duration` and `sampleRate` properties of `AudioBuffer`?

**Comment by guest271314** [comment]
FWIW, from perspective here what we need are tests to substantiate with evidence any adverse impact (performance or otherwise unintended consequences) defining `AudioBuffer` in Worker, and other methods and API's in `AudioWorkletGlobalScope` as currently there is **no _evidence_** that restricting defining Web Audio API methods in Worker or methods available in Worker in `AudioWorkletGlobalScope` will have the adverse (performance) impacts proffered as the reason we cannot just call `fetch()` or `WebTransport` in `AudioWorkletGlobalScope` or use `AudioBuffer` in a `Worker`. We can certainly test this in an "origin trial" or flag to based conclusions on evidence rather than mere conjecture or speculation https://github.com/WebAudio/web-audio-api-v2/issues/110#issuecomment-754883127.

**Comment by guest271314** [comment]
This is one way to exchange data between `Worker` and main thread or `AudioWorkletGlobalScope` 

```
'use strict';
let workerPort;
let buffers;
class WorkletWorker extends AudioWorkletProcessor {
  constructor(options) {
    super(options);
    this.port.onmessage = async e => {
      if (!workerPort) {
        [workerPort] = e.ports;
        const rs = await this.sharedWorkerFetch(
          'data:application/octet-stream;base64,AAAA...'
        );
        const reader = rs.getReader();
        reader
          .read()
          .then(async function processStream({ value, done }) {
            if (done) {
              return await reader.closed;
            }
            globalThis.console.log(new Float32Array(value.buffer));
            return reader.read().then(processStream);
          })
          .then(_ => globalThis.console.log(_));
      }
    };
  }
  async sharedWorkerFetch(url = '', options = {}) {
    if (workerPort) {
      return new ReadableStream({
        start(controller) {
          workerPort.onmessageerror = e => {
            controller.error(e);
          };
          workerPort.onmessage = async ({ data: { value, done } }) => {
            if (done) {
              workerPort.onmessage = null;
              workerPort.onmessageerror = null;
              controller.close();
            } else {
              controller.enqueue(value);
            }
          };

          workerPort.postMessage({ url });
        },
      });
    }
  }
  process(inputs, outputs) {
    return true;
  }
}
registerProcessor('audio-worklet-worker', WorkletWorker);
```

```
'use strict';
let audioWorkletPort;
self.onconnect = e => {
  ([audioWorkletPort] = e.ports);
  audioWorkletPort.onmessage = async e => {
    try {
      const request = await fetch(e.data.url);
      const rs = await request.body;
      const reader = rs.getReader();
      reader
        .read()
        .then(async function processStream({ value, done }) {
          if (done) {
            audioWorkletPort.postMessage({ value, done });
            return await reader.closed;
          } else {
            audioWorkletPort.postMessage({ value, done }, [
              value ? value.buffer : [],
            ]);
          }
          return reader
            .read()
            .then(processStream)
            .catch(e => {
              throw e;
            });
        })
        .catch(e => {
          throw e;
        });
    } catch (e) {
      console.error(e);
      audioWorkletPort.postMessage(e.message);
      throw e;
    };
  };
};
```

The question becomes is the above approach more expensive (N message events) than just using `fetch()` in `AudioWorkletGlobalScope`? Currently there is no way to answer that question conclusively.

**Comment by guest271314** [comment]
Related: FYI WebRTC Insertable Streams has `Worker` tests https://bugs.chromium.org/p/chromium/issues/detail?id=1052765

**Comment by rtoy** [comment]
@chcunningham If the current `AudioBuffer` works for you and the costs that @padenot mentions are acceptable, I have no objections to using the plain `AudioBuffer`.

But this also allows us to update `AudioBuffer` to some form as mentioned in #11 more amenable to the kind of processing you might want to do.

**Comment by padenot** [comment]
> it describes a buffer of raw audio in much the same way as other codec apis (e.g. chrome's internal api, ffmpeg, etc...), including sampleRate, length, duration, channels

I just checked and [Chromium](https://source.chromium.org/chromium/chromium/src/+/master:media/base/audio_buffer.h;l=210;drc=7ccffaf0933ccc647c744bf66971bcf5f33a676a;bpv=1;bpt=1)'s and [ffmpeg](https://ffmpeg.org/doxygen/3.1/group__lavu__sampfmts.html)'s objects specifically supports what I'm talking about here (possibility of having interleaved samples, various bit depths and sample representation in their decoded audio data type). A comment on top of Chrome's declaration also mention that a conversion to an `AudioBus` (the analog object to Web Audio API's`AudioBuffer` in Chromium's codebase) requires copying and a possible conversion (if the source was not f32 planar in the first place).

If we extend `AudioBuffer` to be able to hold interleaved data of different types (as it was requested by authors, but we didn't design a solution yet), then I have of course no objections. I can also take on some of that work (as someone working on both Web Audio and Web Codecs) if we end up finding that it's the right design.

**Comment by guest271314** [comment]
From the deletion of comments as "off-topic" I can only conclude that Web Audio API does not rely on empirical evidence for conclusions. The comments are squarely _on-topic_.

**Comment by guest271314** [comment]
Why delete comments critical of decisions based on conjecture without evidence? 

Is this 

> FWIW, from perspective here what we need are tests to substantiate with evidence any adverse impact (performance or otherwise unintended consequences) defining AudioBuffer in Worker, and other methods and API's in AudioWorkletGlobalScope as currently there is no evidence that restricting defining Web Audio API methods in Worker or methods available in Worker in AudioWorkletGlobalScope will have the adverse (performance)

not true and correct?

If so, where are the tests?

**Comment by rtoy** [comment]
Comments were not deleted. I can still see them if I click them.

**Comment by hoch** [comment]
@padenot Thanks for the context! As a someone who was not involved in WebCodec's design, I am curious. Then why does AudioFrame include AudioBuffer? I believe AudioBuffer is specifically designed for Web Audio, so non-interleaved format only makes sense.

**Comment by chcunningham** [comment]
> Then why does AudioFrame include AudioBuffer?

The full reasoning is in [my comment above](https://github.com/WebAudio/web-audio-api-v2/issues/111#issuecomment-756541266). I made the call pretty much independently without consideration of the tradeoffs we're now discussing. Still, I think those points in favor of using AudioBuffer are compelling and it sounds like we have a solid path forward to resolve the issues raised here. @hoch do you agree? 

> I believe AudioBuffer is specifically designed for Web Audio, so non-interleaved format only makes sense.

If WebCodecs were to create its own AudioBuffer, it would be pretty much identical to WebAudio's except, perhaps, offering more formats (and clearly that is not a v1 requirement, even for WebCodecs). If we can instead add those formats to AudioBuffer I think it simplifies how raw audio is described on the web platform and improves interop with WebAudio.

**Comment by hoch** [comment]
I discussed the GC pressure issue with @chcunningham and we believe Chromium's GC can handle the pressure from the rapid AudioBuffer object creation from the WebCodecs API. @padenot might be able to speak for FireFox's view on this.

Here I see two paths:
1. Expose BAC to Workers: By exposing BAC, we have to open other related interfaces to Workers anyway. This is a longer-term project.
2. Expose AudioBuffer to Workers first, and then BAC later: We can support WebCodecs' use case by opening AudioBuffer in Workers. The change seems straightforward and it will be a great pilot project for the BAC in the future.

The option 2 seems non-controversial. I am curious what other people think.

> offering more formats (and clearly that is not a v1 requirement, even for WebCodecs)

This deserves a separate issue in this tracker. Definitely not a V1 or short-term project.

**Comment by chcunningham** [comment]
I agree option 2 seems not controversial. Particularly the firs step of exposing AudioBuffer to workers. I'm happy to see the other parts of option 2 pursued as well, but for now I'd like to just unblock on changing the exposure. 

@padenot @rtoy @hoch are we in agreement?

**Comment by hoch** [comment]
Yes. I agree.

**Comment by hoch** [comment]
@padenot I understand we generally agree on this. Right?

**Comment by guest271314** [comment]
Given user-defined `Float32Array` set to `AudioBuffer` how to generate a `timestamp` (https://bugs.chromium.org/p/chromium/issues/detail?id=1185755) for `AudioFrame`?

**Comment by guest271314** [comment]
I have not found WebCodecs 'opus' `AudioEncoder` and `AudioDecoder` to be reliable, neither real-time, nor storing the value and streaming only when the entire input has been decoded. 

Nonetheless, given `AudioFrame` is used by Media Capture Insertable Streams (Breakout Box; Media Transform API), we still face the issue of generating `timestamp` programmatically.

One way to generate `timestamp`s is to create a `MediaStreamAudioDestination` connected to an `OscillatorNode` (Chrome implementation of `MediaStreamAudioDestinationNode` is not in compliance with the specification; is not "potentially playing" although `readyState` is `"live"`; requires an `AudioNode` as source to commence playback at `HTMLMediaElement`, else we would not need to create an `OscillatorNode`) with `frquency` set to `0` and use the `timestamp` from the `AudioFrame` from `value` at `ReadableStreamDefaultReader,read()` to supply user-defined `buffer` and `timestamp` at new `AudioFrame` passed to `WritableStreamDefaultWriter.write()`, something like

           const { stream } = msd;
           const [track] = stream.getAudioTracks();
           const osc = new OscillatorNode(ac, { frequency: 0 });
           const processor = new MediaStreamTrackProcessor(track);
           const generator = new MediaStreamTrackGenerator('audio');
           const { writable } = generator;
           const { readable: audioReadable } = processor;
           const audioWriter = writable.getWriter();
           const mediaStream = new MediaStream([generator]);
           const audioReader = audioReadable.getReader();
           audio.srcObject = mediaStream;
           osc.connect(msd);
           osc.start();
           let duration = 0;
            audioReader.read().then(async function process({ value, done }) {
              // avoid glitches, gaps, clipping start of MediaStreamTrackGenerator output (~0.5 second)
              if (audio.currentTime < value.buffer.duration * 50) {
                return audioWriter
                  .write(value)
                  .then(() => audioReader.read().then(process));
              }
              if (writeOffset && writeOffset >= readOffset) {
                // avoid clipping end of MediaStreamTrackGenerator output (~1.0 seconds; await audio playback completion)
                if (audio.currentTime < duration + value.buffer.duration * 100) {     
                  // noop           
                  return audioReader.read().then(process);
                } else {
                  msd.disconnect();
                  osc.disconnect();
                  track.stop();
                  audioReader.releaseLock();
                  audioReadable.cancel();
                  audioWriter.releaseLock();
                  writable.abort();
                  await writable.closed;
                  await ac.close();
                  console.log(`readOffset: ${readOffset}, writeOffset: ${writeOffset}, duration: ${duration}, audio.currentTime: ${audio.currentTime}, ac.currentTime: ${ac.currentTime}`);                
                  return await Promise.all([
                    new Promise((resolve) => (stream.oninactive = resolve)),
                    new Promise((resolve) => (ac.onstatechange = resolve)),
                  ]);
                }
              }
              // get timestamp from AudioFrame
              const { timestamp } = value;
              // ...
              const floats = new Float32Array(220);
              // set user-defined values at floats
              const buffer = new AudioBuffer({
                numberOfChannels: 1,
                length: floats.length,
                sampleRate: 22050,
              });
              buffer.copyToChannel(floats, 0, 0);
              duration += buffer.duration;
              const frame = new AudioFrame({ timestamp, buffer });
              return audioWriter.write(frame).then(() => {
                frame.close();
                return audioReader.read().then(process);
              });
            })

**Comment by guest271314** [comment]
Exposing `AudoBuffer` is not an issue. We can effectively create a user-defined audio buffer using `Float32Array`.

>  In both cases, we wrap the buffer in an [AudioFrame](https://wicg.github.io/web-codecs/#audioframe-interface) to add some metadata and the ability to immediately release the buffer without waiting for GC

The issue is the wrapping in an `AudioFrame` which has the `timestamp` property that has serious problems in production (I am not sure how `AudioEncoder` and `AudioDecoder` have been deployed, because they do not work, are broken https://plnkr.co/plunk/yY4CZbPNnZMT7Kwd; perhaps save for a test with a single frame).

An exposed `AudioBuffer` in a `Worker` will still be less than useful when the `timestamp` attribute of the wrapping `AudioFrame` can be backwards in time with respect to the preceding `AudioFrame`(s)

```
AudioFrame {timestamp: 54858, buffer: AudioBuffer}
AudioFrame {timestamp: 54293, buffer: AudioBuffer}

AudioFrame {timestamp: 126129, buffer: AudioBuffer}
AudioFrame {timestamp: 125602, buffer: AudioBuffer}

1194760315999
1194760315470

1194760347008
1194760346476
```
Fixing that should be the priority.

**Comment by guest271314** [comment]
WICG banned me. Otherwise I would strongly suggest filing a PR in WebCodecs specification that includes the language that `timestamp` MUST be greater (which includes non-duplication) than the previous `timestamp` https://bugs.chromium.org/p/chromium/issues/detail?id=1181547.

**Comment by guest271314** [comment]
> WebCodecs uses AudioBuffer as a container of unencoded audio.
https://wicg.github.io/web-codecs/#dom-audioframe-buffer

I am curious why `AudioFrame` is necessary at all?

What is the reasoning for using a `timestamp` attribute?

`AudioWorklet` works without a `timestamp`.

`AudioFrame` (and the "microsecond" `timestamp` that is used internally for Chrome implementation, though not exposed as constructable to the Web) appears to be an unnecessary abstraction.

**Comment by chcunningham** [comment]
Looks like the build is happy now!

I've updated the opening comment w/ links to chromium bug and WPT tests. Please let me know if anything else is needed.

**Comment by padenot** [comment]
@chcunningham this will happen automatically when we expose the Web Audio API in Web Workers, and Web Codecs doesn't need it anymore, so I'm closing this, thanks!
